ported to use fetch; streamlined all imports to be es6 imports

damooo opened the pull request:
this PR performs following:

- uses fetch for network requests. as mentioned in #55 
- cleared any commonjs `require()` imports, all imports are now ES6 imports.. this case happened with previous `http`, etc modules and `@iiif/vocabulary`. now first case addresed by above, second case now imports only through `@iiif/vocabulary/dist-commonjs` in es6 import syntax. thus previous two imports  like, es6`ServiceProfile` and commonjs`ServiceProfileEnum` imports are merged to one es6`ServiceProfile` import as both refers to same binding.

### results

- now umd bundle size is 40% that of previous bundle. : ) . ~3300 LOC compared to ~13000 LOC previously without minification of bundle.
- mocha tests are passed in node environment. worked normal in browser.

damooo commented:
presently it uses `vedavaapi-web` package for wrapping web apis like node-fetch and browser-fetch with compatability considerations. we can make it available from iiif-commons, as iiif-web or something if required.

mejackreed commented:
👏 Thanks for moving this forward @damooo . I'll try and take this for a spin in Mirador 3. We presently also rely native fetch so this seems like a great improvement. 

I'm curious about the impact (if any) on IIIF auth services so want to take a deeper look.

damooo commented:
@mejackreed it uses fetch in browser, and nothing else. so there should not be any problem in broweser.
on node, `node-fetch` doesn't supports cookies by default. above mentioned package wraps node-fetch with a `tough-cookie` cookie-jar annotation, and hence supports cookies in standard way, along with `node-fetch` capabilities. hence there should be no problem.
here is code for above package https://gitlab.com/damodara/vedavaapi-web/. it is 3 tiny modules.

damooo commented:
@mejackreed  a note that, if we import manifesto, in app like mirador, we have to add `node-fetch`, `url`, etc.. some four packages to `webpackConf.external`, so that webpack doesn't include them in umd bundle. you can refer to this commit's webpack.config.js. here is the link https://github.com/damooo/manifesto/blob/f4ec06041c86da6f106eae4c60f5a38b90849b10/webpack.config.js#L14
same goes for rollup.

stephenwf commented:
@damooo would it be possible to [change](https://unpkg.com/browse/vedavaapi-web@0.0.5/package.json) the `dependencies` to `peerDependencies` so that we could bump versions of `node-fetch`, `fetch-cookie` and `form-data` without waiting for a new release of `vedavaapi-web`. 

Also is this library a no-op in regular bundles, or does it include a polyfill for fetch?

damooo commented:
yes. i will make them peer dependencies.
it doesn't include any polyfills.. it just proxies bindings to web apis like `fetch`, '`URL`, 'URLSearchParams`, `FormData`, etc.. properly, in browser to browser natives, and in node to respective equivalents. hence shouldn't have any problem. this takes some 15 LOC.

along with them it provides one other function `request`, it takes same params as `fetch`, and handles errors properly. it rejects promise for 40x, 50x, etc.. along with attaching error body, context. it is some 30 LOC

only note for bundling is as mentioned above... we have to set `node-fetch`, `url`, `fetch-cookie`, `form-data` as externals in bundle config, for umd builds... so that bundles wouldn't include them. for webpack, we can refer this commit's config.

damooo commented:
@stephenwf updated package with `dependencies` => `peerDependencies` change.

edsilv commented:
Cool. Everyone happy to merge this?

damooo commented:
@edsilv travis is failing if we make them peerDependencies. how to tell travis to install peerDeps?

mejackreed commented:
I've added them as `devDependencies` also in some cases, especially if travis requires them for it to work correctly.

damooo commented:
finally... i added npm install commands to travis before_install hook, so that it can install peer dependencies, as mentioned in a SO answer.

stephenwf commented on webpack.config.js:
```diff
@@ -10,6 +10,13 @@ const config = {
+        'node-fetch': 'nodeFetch',
+        'fetch-cookie/node-fetch': 'fetchCookieWrapper',
+        'form-data': 'formData',
+        'url': 'nodeUrl',
```
I think for externals, the value on the right needs to be either: 
* Global from a UMD module (e.g. jQuery's `$`)
* AMD-style include (e.g. a package name `node-fetch`)

Couple of examples in the webpack repo: https://github.com/webpack/webpack/tree/master/examples/externals

I think if you changed them to:
```
{
        'node-fetch': 'node-fetch',
        'fetch-cookie: 'fetch-cookie',
        'form-data': 'form-data',
        'url': 'url',
}
```
then it would work, although I'm not exactly sure why they are being included in the bundle. I feel like this might be getting more complicated then a simple `node-fetch` shim for testing (included in the bootstrap for the tests). 

That would leave it up to users of Manifesto to decide if they want a browser shim for fetch or not.

damooo replied:
in above conf, right side global binding names are not valid js binding names, and build will reject it. i will check once though.

damooo replied:
@stephenwf it worked!

damooo commented on test/tests/deep-hierarchy.js:
```diff
@@ -7,6 +7,7 @@ var manifest, ceili;
 
 describe('#loadsTop', function() {
     it('loads successfully', function (done) {
+        console.log(manifests.deephierarchytop);
```
@mejackreed i will remove that console.log, can you share mirador branch once, so that i can check it once..
and for importing from code, we can import directly from manifest.js, as pkg.main is setted to dist-commonjs/index.js . umd bundle can be used to directly embed manifesto in links explicitly.

mejackreed replied:
Here is the Mirador branch, however I switched all to es6 imports (one just needs to be converted over). Also local manifesto path in `package.json` might need updating: https://github.com/ProjectMirador/mirador/compare/updated-manifest

damooo commented on .travis.yml:
```diff
@@ -5,4 +5,8 @@ before_script:
\ No newline at end of file
+  - npm test
+before_install:
+  - npm install node-fetch
```
so which version constrain we can give for it, so that it allows updating node-fetch with out restrain? i will try with `"node-fetch": "*"`, so that dev dependencies pull version of user constraint specific

damooo replied:
now this is addressed

damooo commented:
@mejackreed  @stephenwf @edsilv here is status as of this commit:

## dependency installation:
- dependencies `node-fetch`, `fetch-cookie`, `form-data` are now added to devDependencies, with version being `*`, so that top level apps can choose their version i they want, or retrieves latest version by `npm install` with out further headache of installing peer deps
- travis config resetted to previous after above change

## build config:
as suggested above, now modified global export name to same as module name for `externals` this solves above problem

## using in build:
- presently build generates three packages `dist-commonjs`, `dist-esmodule`, `dist-umd`.
- `pkg.main` is `dist-commonjs/index.js`; hence when we import by `require('manifest.js')` it will use `dist-commonjs/index.js`,  in webpack, we can import using es6 import too natively, rollup needs standard rollup-plugin-commonjs plugin for build.
- `dist-esmodule` can be directly imported using es6 modules if needed.
- `dist-umd/manifest` works in browser using `source` tag. it is mainly meant for browser environment. and now it also works in node if required. but it is recommended to import from pkg.main directly, as it can elemenate dead code, and no unnecessary deps will be include.
- use this present commit's webpack config in apps for externals.
```js
externals: {
        'node-fetch': 'node-fetch',
        'fetch-cookie/node-fetch': 'fetch-cookie/node-fetch',
        'form-data': 'form-data',
        'url': 'url',
    }
```
